disco_aws_automation/disco_elasticsearch_archive.py

```diff
@@ -0,0 +1,498 @@
+                credentials.access_key,
+                credentials.secret_key,
+                self.region,
+                'es'
+            )
+            self._es_client = Elasticsearch(
+                [self.host],
+                http_auth=aws_auth,
+                use_ssl=False,
+                verify_certs=False,
+                connection_class=RequestsHttpConnection,
+            )
+        return self._es_client
+
+    @property
+    def disco_es(self):
+        if not self._disco_es:
+            self._disco_es = DiscoElasticsearch(self.environment_name)
+
+        return self._disco_es
+
+    @property
+    def disco_iam(self):
+        if not self._disco_iam:
+            # TODO: the environment argument for DiscoIAM is actually the account
+            # name, e.g. "dev" or "prod"
+            self._disco_iam = DiscoIAM(environment=self.environment_name)
+
+        return self._disco_iam
+
+    @property
+    def s3_client(self):
+        if not self._s3_client:
+            self._s3_client = boto3.client("s3")
+
+        return self._s3_client
+
+    @property
+    def s3_bucket_name(self):
+        if not self._s3_bucket_name:
+            self._s3_bucket_name = "{0}.es-{1}-archive.{2}".format(
+                self.region, self.cluster_name,
+                self.environment_name)
+
+        return self._s3_bucket_name
+
+    def _bytes_to_free(self, threshold):
+        """
+        Number of bytes disk space threshold is exceeded by. In other
+        words, number of bytes of data that needs to be delted to drop
+        below threshold.
+        threshold range 0-1
+        """
+        cluster_stats = self.es_client.cluster.stats()
+        total_bytes = cluster_stats["nodes"]["fs"]["total_in_bytes"]
+        free_bytes = cluster_stats["nodes"]["fs"]['free_in_bytes']
+        need_bytes = total_bytes * (1 - threshold)
+        logging.debug("disk utilization : %i%%", (total_bytes-free_bytes)*101/total_bytes)
+        return need_bytes - free_bytes
+
+    def _indices_to_delete(self, threshold):
+        """
+        Return list of indices that have already been archived, so that when they are deleted,
+        disk space would get below the threshold. Oldest first.
+        """
+        bytes_to_free = self._bytes_to_free(threshold)
+        logging.debug("Need to free %i bytes.", bytes_to_free)
+        if bytes_to_free <= 0:
+            return []
+
+        index_sizes = self._get_all_indices_and_sizes()
+        indices_to_delete = []
+        freed_size = 0
+        for index, size in index_sizes:
+            if self.snapshot_state(index) == 'SUCCESS':
+                indices_to_delete.append(index)
+                freed_size += size
+                if freed_size >= bytes_to_free:
+                    break
+        return indices_to_delete
+
+    def _archivable_indices(self):
+        """ Return list of indices that are older than today's date """
+        # TODO: might want to consider timezone issue?
```

Comment:
We could use yesterday 24/48 hours doesn't make much of a difference but does skirt around timezone issues. Another option is to find latest index and only skip archiving it.

All suggestions NTH.